### PR TITLE
optimization of U256

### PR DESCRIPTION
**Pull request by NikVolf**
before
```
test u256_add ... bench:     139,774 ns/iter (+/- 8,128)
test u256_mul ... bench:   4,915,555 ns/iter (+/- 436,083)
test u256_sub ... bench:     757,326 ns/iter (+/- 47,169)

```

after
```
test u256_add ... bench:      36,160 ns/iter (+/- 2,217)
test u256_mul ... bench:     104,526 ns/iter (+/- 3,128)
test u256_sub ... bench:      35,494 ns/iter (+/- 1,163)
```

### CODE REVIEW

**Comment by debris** [comment]
nice, benchmarks are really impressive ^^ :+1:

**Comment by NikVolf** [comment]
just good old asm :)

**Comment by arkpar** [comment]
What about U512? That's even more of a bottleneck currently.
I think we don't actually need multiplication of 512 -bit uints, We can get away with a custom function that multiplies two U256s into a U512.

File: util/benches/bigint.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+extern crate ethcore_util;
+
+use test::{Bencher, black_box};
+use ethcore_util::uint::*;
+
+#[bench]
+fn u256_add(b: &mut Bencher) {
+	b.iter(|| {
+		let n = black_box(10000);
+		(0..n).fold(U256::from(1234599u64), |old, new| { old.overflowing_add(U256::from(new)).0 })
```
Comment (arkpar) [inline util/benches/bigint.rs]: benchmark is not very accurate having a call to `U256::from` on each iteration.

File: util/src/lib.rs [hunk 2]
```diff
@@ -16,6 +16,7 @@
 
 #![warn(missing_docs)]
 #![cfg_attr(feature="dev", feature(plugin))]
+#![cfg_attr(feature="dev", feature(asm))]
```
Comment (arkpar) [inline util/src/lib.rs]: The `dev` feature is for dev dependencies like clippy. Something else should be used here

File: util/src/uint.rs [hunk 3]
```diff
@@ -51,6 +51,232 @@ macro_rules! impl_map_from {
...
+macro_rules! uint_overflowing_add {
+	(U256, $n_words: expr, $self_expr: expr, $other: expr) => ({
+		let mut result: [u64; 4] = unsafe { mem::uninitialized() };
+		let self_t: &[u64; 4] = unsafe { &mem::transmute($self_expr) };
+		let other_t: &[u64; 4] = unsafe { &mem::transmute($other) };
+
+		let overflow: u8;
+        unsafe {
+            asm!("
+                adc $9, %r8
```
Comment (arkpar) [inline util/src/uint.rs]: why not use auto-assigned registers?

File: util/src/uint.rs [hunk 4]
```diff
@@ -51,6 +51,232 @@ macro_rules! impl_map_from {
...
+                adc $10, %r9
+                adc $11, %r10
+                adc $12, %r11
+                setc %al
+                "
+            : "={r8}"(result[0]), "={r9}"(result[1]), "={r10}"(result[2]), "={r11}"(result[3]), "={al}"(overflow)
+            : "{r8}"(self_t[0]), "{r9}"(self_t[1]), "{r10}"(self_t[2]), "{r11}"(self_t[3]),
+			  "m"(other_t[0]), "m"(other_t[1]), "m"(other_t[2]), "m"(other_t[3])
```
Comment (arkpar) [inline util/src/uint.rs]: Input should be allowed to be a register in case it is inlined and compiler already has the value in the register

**Comment by arkpar** [comment]
`Sub::sub` is still not using the asm version?

**Comment by NikVolf** [comment]
@arkpar `sub` uses `overflowing_add`
so it should become better as well

@arkpar 
only u256 in this pr
u512 comes next

@arkpar 
other issues resolved

**Comment by arkpar** [comment]
build warnings https://travis-ci.org/ethcore/parity/jobs/111836764#L824

**Comment by debris** [comment]
You can also move assembly code to separate `.s` files and compile it with gcc. It would make optimization available also on rust beta/stable.

**Comment by arkpar** [comment]
@marek We want the calls to be inlined. Also less less problems with windows later

**Comment by NikVolf** [comment]
removed warning (cause it's macros expansion, had to ignore it)

File: util/src/uint.rs [hunk 5]
```diff
@@ -51,6 +51,233 @@ macro_rules! impl_map_from {
...
+#[cfg(not(all(feature="x64asm", target_arch = "x86_64")))]
+macro_rules! uint_overflowing_sub {
+	($name:ident, $n_words: expr, $self_expr: expr, $other: expr) => ({
+		let res = overflowing!((!$other).overflowing_add(From::from(1u64)));
+		let res = overflowing!($self_expr.overflowing_add(res));
+		(res, $self_expr < $other)
+	})
+}
+
+#[cfg(all(feature="x64asm", target_arch = "x86_64"))]
```
Comment (gavofyork) [inline util/src/uint.rs]: space around `=`.

File: util/src/uint.rs [hunk 6]
```diff
@@ -51,6 +51,233 @@ macro_rules! impl_map_from {
...
+		(U256(result), overflow != 0)
+	});
+	($name:ident, $n_words: expr, $self_expr: expr, $other: expr) => ({
+		let res = overflowing!((!$other).overflowing_add(From::from(1u64)));
+		let res = overflowing!($self_expr.overflowing_add(res));
+		(res, $self_expr < $other)
+	})
+}
+
+#[cfg(all(feature="x64asm", target_arch = "x86_64"))]
```
Comment (gavofyork) [inline util/src/uint.rs]: space around `=`.

File: util/src/uint.rs [hunk 7]
```diff
@@ -51,6 +51,233 @@ macro_rules! impl_map_from {
...
+			let ret = overflowing!($name(ret).overflowing_add($name(carry)), overflow);
+			(ret, overflow)
+		} else {
+			($name(ret), overflow)
+		}
+	})
+}
+
+
+#[cfg(all(feature="x64asm", target_arch = "x86_64"))]
```
Comment (gavofyork) [inline util/src/uint.rs]: space around `=`. every other time in here, too.

File: util/src/uint.rs [hunk 8]
```diff
@@ -51,6 +51,233 @@ macro_rules! impl_map_from {
...
+				adc $$0, %rdx
+				or %rdx, %rcx
+
+				mov $5, %rax
+				mulq $12
+				add %rax, $3
+				adc $$0, %rdx
+				or %rdx, %rcx
+
+                cmpq $$0, %rcx
```
Comment (gavofyork) [inline util/src/uint.rs]: we want tabs!

**Comment by gavofyork** [comment]
minor style stuff. can't speak for the logic and i don't see any additional tests... coverage seems to have gone down slightly - is this spurious or an indication that additional code paths need testing?

**Comment by NikVolf** [comment]
@gavofyork 
yeah right
i have plenty of tests here: https://github.com/NikVolf/asm-fun/blob/master/src/lib.rs
will migrate it

**Comment by NikVolf** [comment]
@debris indeed, Arkady right, inlining gives 2x boost to performance for add/sub

**Pull request merged**